### PR TITLE
challenge: Subtract two numbers

### PR DESCRIPTION
**Pull request by GeehanAli**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by misik-eng — COMMENTED** [review]
### Files

- [X] The file name describes the function's behavior
- [X] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [ ] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [X] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

## Strategy

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

File: solutions/tests/test_subtract_two_numbers.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# python3
+# coding
+"""
+Test module for subtract_two_numbers
```
Comment (misik-eng) [inline solutions/tests/test_subtract_two_numbers.py:4]: Maybe there could be more detailed docstring for test module. But it looks good.

File: solutions/tests/test_subtract_two_numbers.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import unittest
+
+from ..subtract_two_numbers import subtract_two_numbers
+
+
+class TestSubtractTwoNumbers(unittest.TestCase):
+    """Test suit to subtract_two_numbers function"""
+
+    def test_subtract_two_numbers_positive_int(self):
```
Comment (misik-eng) [inline solutions/tests/test_subtract_two_numbers.py:18]: Every unit test should has a clear docstring.

!!!

File: solutions/subtract_two_numbers.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    Raises:
+        TypeError: If either a or b is not a number (int or float).
+
+    """
+    assert isinstance(a, (int, float)), "First argument must be a number (int or float)"
+    assert isinstance(
+        b, (int, float)
+    ), "Second argument must be a number (int or float)"
+
```
Comment (misik-eng) [inline solutions/subtract_two_numbers.py:44]: There could be a comment about the strategy, but it is easy enough to understand, so it is not necessary.

File: solutions/subtract_two_numbers.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# python3
+# coding
+"""creating on 01 03 2025
+@author Geehan Ali + ChatGPT(phind)"""
+
+from typing import Union
+
+
+def subtract_two_numbers(
+    a: Union[int, float], b: Union[int, float]
```
Comment (misik-eng) [inline solutions/subtract_two_numbers.py:10]: There might be clearer variable names instead of 'a' and 'b,' but the strategy is simple enough to understand what those variables represent, so I guess it’s okay.

File: solutions/tests/test_subtract_two_numbers.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        self.assertEqual(subtract_two_numbers(5, 4), 1)
+
+    def test_Subtract_Two_Numbers_int_float(self):
```
Comment (misik-eng) [inline solutions/tests/test_subtract_two_numbers.py:21]: Every unit test should has a clear docstring.

!!!

**Review by misik-eng — COMMENTED** [review]
I checked our checklist, and your code is almost perfect, but it should have a clear docstring for each test. Otherwise, it is crystal clear.

File: solutions/tests/test_subtract_two_numbers.py (hunk 1 above)
Comment (GeehanAli) [reply solutions/tests/test_subtract_two_numbers.py:4]: Absolutely right.
Done!
